File: backend/src/middleware/token.ts
import { Request, Response, NextFunction } from 'express';
import { verify } from '../util/token';

export async function tokenSession(request: Request, _response: Response, next: NextFunction): Promise<void> {
    let token: string = request.body.token || request.query.token || request.headers['x-access-token'];
    if(!token || token == 'null') {
        next();
        return;
    }

    let data = await verify(token);
    request.session.token = token;
    request.session.tokenData = data;
    next();
}

File: backend/middleware/token.go
package middleware

import (
	"net/http"

	"github.com/winded/tyomaa/backend/util"
	"github.com/winded/tyomaa/backend/util/context"
	"github.com/winded/tyomaa/backend/util/token"
	"github.com/winded/tyomaa/shared/api"
)

func TokenSession(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		tk := r.Header.Get("X-Access-Token")
		if tk == "" {
			next.ServeHTTP(w, r)
			return
		}

		ctx := context.Get(r)

		claims, err := token.Verify(tk)
		if err != nil {
			util.WriteApiError(w, api.Error(http.StatusBadRequest, "Invalid or malformed token"))
			return
		}

		ctx.Token = tk
		ctx.Claims = claims

		r = context.Set(r, ctx)
		next.ServeHTTP(w, r)
	})
}
